### ISSUE TITLE
Smearing and CTI

### ISSUE THREAD

**Issue by drsamadi** [IvS-KULeuven/PlatoSim3#692]
Hi,
I'm looking at the smearing (the data stored in  SmearingMaps) and get such values

![Screenshot from 2022-04-15 16-33-25](https://user-images.githubusercontent.com/19569211/163584141-7dce2d20-1f67-49aa-b557-5ab9be99e0f0.png)

black: no CTI
red: with CTI

SmearingMaps is plotted here as a function of the row index (for a particular CCD column). They are 300 over-scan rows simulated. Unit is e-.

 I don't understand  why when CTI is activated  the signal is low at the beginning and then increases. I'm  expecting the reverse: high at the begging and then decreasing with the row index, which is that I get with PIS (see PLATO-LESIA-PDC-TN-0030, i0.2 ). 
Indeed, the over-scan rows (virtual pixel)  just follow the physical lines of the CCD that are transferred before the over-scan rows. 
The physical pixels have already gone through the traps and filled them. The virtual pixels should then rather collect the charges released by traps already filled by physical pixels during the transfer. Accordingly, the virtual pixels collect additional charges and therefore the signal in the over-scan rows should  - in the presence of CTI -  be higher than the level of the smearing signal that one gets in the absence of CTI.

Réza

**Comment by Dries-Seynaeve** [comment]
Hi Réza,

I agree this shouldn't be the case. Can you share the code you used to obtain the image together with the input file? I will take a look at what the problem might be.  

Kind regards
Dries

**Comment by drsamadi** [comment]
[00.yaml.txt](https://github.com/IvS-KULeuven/PlatoSim3/files/8496644/00.yaml.txt)
[01.yaml.txt](https://github.com/IvS-KULeuven/PlatoSim3/files/8496645/01.yaml.txt)



a0=h5py.File('00/00.hdf5','r')
a1=h5py.File('01/01.hdf5','r')
s1 = np.array(a1['SmearingMaps/smearingMap000000'])
s0 = np.array(a0['SmearingMaps/smearingMap000000'])

o=1e3
g=25

plot((s0[:,1500]-o)*g,'k')
plot((s1[:,1500]-o)*g,'r')

**Comment by Dries-Seynaeve** [comment]
Hello Réza, 

The problem you describe was caused by how we set the initial value of our occupied traps. This was set to 0, so the effect you see is because not enough traps are filled up when we read out the smearing map, and we lose flux (compared to without CTI). 

I changed the initial value of these occupied traps, and also changed the number of occupied traps when we read out the smearing map. Before I push these changes, could you take a look at this figure and see if the order of magnitude difference between CTI and no CTI makes sense. 

The figure is similar to the one you showed but I also showed a cross-section of the subfield (without any stars exposed to it). The smearing map shows the effect that you described in your issue, I just want to make sure that what we obtain is realistic. 

Kind regards 
Dries 
![pixel-smearingMapCTI](https://user-images.githubusercontent.com/74176898/173373764-00cc2b32-436a-402d-afbf-e1f48dee89bb.png)

**Comment by drsamadi** [comment]
Hi Dries,
looking at your figures it seams that the problem is now fixed. I'm going to test it soon. 
Réza

**Comment by drsamadi** [comment]
Do you have committed your change ? I have tested the latest version (PlatoSim 3.5.3-40-g19f3cb4d) and this version shows the same problem.

**Comment by Dries-Seynaeve** [comment]
Hi Reza,

I haven’t pushed this yet, I am currently on vacation but I will push it Monday when I’m back.

Kind regards
Dries

**Comment by Dries-Seynaeve** [comment]
Hi Réza

I have pushed the changes. 

Kind regards 
Dries

**Comment by drsamadi** [comment]
Hi Dries,

I have checked the new version.
Very qualitatively this is fine, however, using the same CTI parameters I get very different results with my own implementation of Short et al (2013).

This is illustrated in the two plots below:

![Screenshot from 2022-07-25 12-47-30](https://user-images.githubusercontent.com/19569211/180760266-6b3f36c1-82dc-4e44-821e-19ddd2be96b0.png)
![Screenshot from 2022-07-25 12-47-42](https://user-images.githubusercontent.com/19569211/180760270-3cbb3f27-6f21-4c1b-9531-36b66b871ebc.png)

black: no CTI
blue : CTI (PLATOSim)
green: CTI (my implementation)

The  two figures show the (parrallel) over-scan pixel values as a function of the row index (taken from SmearingMaps). The first plot correspond the first CCD column and the second plot to the column  of index 100.

We see that with PLATOSim the trail induced by the CTI decreases much quickly than in my calculation.
I see also that  with my calculation the CTI starts with higher values (in the figure unit is e-).  

Réza
PS: Note that my calculation does not shows discontinuities since I do not digitalise the pixel values  
(i.e. I do not apply the quantification of the signal).

**Comment by Dries-Seynaeve** [comment]
Hi Réza

I suspect the problem lies here in how the occupied traps are set for the smearingmap. If this is the issue, then I see two immediate places where I could have made a mistake. 

1. when setting the number of occupied traps for the pixelmap. I don't think this is the issue, but you can check this to make sure by making the same plot as above for the subfield. If we have different pixel values we should see this in that figure. 

2. when I estimate the number of occupied traps for the smearingmap.  To estimate this, I assumed a uniform distribution of (occupied) traps and estimated the number of occupied traps for the smearingmap as (4510 / lastRowPixelMap) * numberOccupiedTrapsPixelMap. I might have made a mistake in my assumption. 

So could you do a similar plot for the pixelmap as you did for the smearingmap? If the same issue occurs, the problem might lie in how the inital occupied traps for the pixelmap are calculated, otherwise the issue might be in how they are calculated for the smearing map. 

Kind regards 
Dries

**Comment by drsamadi** [comment]
Hi Dries,
here a plot showing for the first column the last 100 lines (in the CCD area -> "pixelmap"):

![Screenshot from 2022-07-26 14-15-26](https://user-images.githubusercontent.com/19569211/181003562-9f4d3a2c-e5f1-4b0e-8261-c4bdd54eacf9.png)

both calculations do not exactly match but it seams than they are qualitatively closer.

Réza

**Comment by Dries-Seynaeve** [comment]
Hi Reza, 

That is good, this seems to suggest that the issue is related to how the number of occupied traps for the smearing is calculated. For PlatoSim with the subfield as defined in your yaml files, I believe the number of occupied traps that the first row of the smeeringmap will see is the same as the last number of occupied traps that is obtained after reading out the last line of the pixelmap. So for the CTI effect, the simulator simply sees this as if the smearingmap + subfield is one CCD. 

If there isn't something glaringly wrong with this assumption, would you be willing to share the part of your code that deals with the CTI? Maybe I can go through it and see what we do differently. 

Kind regards 
Dries

**Comment by drsamadi** [comment]
yes this is correct, the over-scan area and the CCD area can be treated as a single "CCD"
The Short et al's model requires that you fixe the "dwell" time, what is the value you are using in platosim ? how do you compute it ?
Here below the piece of code I'm using to compute it:
serial_transfers = 2255
serial_prescan = 25
serial_overscan = 220
readout_freq = 3e6
phase_overlap = 5e-6
pt = 22.5e-6
rise_time = 25e-6
shift_duration = 4*phase_overlap+pt+4*rise_time
dwell_time = (serial_transfers+serial_prescan+serial_overscan)/readout_freq
ptp = shift_duration+dwell_time
print('readout freq. = %.2f MHz, ptp = %.2f ms, parallel shift duration = %.2f ms'%(readout_freq/1e6, ptp*1e3, shift_duration*1e3))
print('dwell_time = %f ms' % (dwell_time*1E3))

**Comment by Dries-Seynaeve** [comment]
PlatoSim uses the parallelTransferTime parameter (given in the input file, under the CCD section) as "dwell time".

**Comment by drsamadi** [comment]
this is a very short time-scale ; my dwell time is 833 micro seconds while parallelTransferTime is 110 micro seconds in the YAML file.  This parallelTransferTime (= parallel shift duration  in my calculation above) is the time needs to move one row however, the dwell time is the time during which the register is read out, as far I've understood this is the relevant time scale for the CTI.

**Comment by Dries-Seynaeve** [comment]
Ok, that makes sense. I will change this in PlatoSim.

**Comment by drsamadi** [comment]
I have tested that I get when the time scale is 110 micro seconds:

![Screenshot from 2022-07-27 09-53-37](https://user-images.githubusercontent.com/19569211/181193009-b51b3eb0-9e6d-4c70-a86f-66447c1c70fd.png)


black: 110 micro seconds (this is the transfer time)
red: 833 micron seconds (this is the "dwell time")

Of course if you change the  time scale in platosim this will impact all the simulations done so far for instance those generated recently by @nicholasjannsen

**Comment by Dries-Seynaeve** [comment]
Hi Reza, 

I have changed how we calculate the dwell time. This now seems to agree more with your results. 
We do still have a different dwell time, this is because PlatoSim does not include serial overscan rows. I seem to get a dwell time of 775,2 microseconds, which seems to be the right order of magnitude. (if we included these overscan rows we would have a dwell time of 850 microseconds)

Kind regards 
Dries 

![Figure_1_dt](https://user-images.githubusercontent.com/74176898/181212017-48b6c4a3-1c29-4594-89a8-8cc4d789e738.png)

**Comment by drsamadi** [comment]
OK I understand you calculation of the dwell time should be consistent with the assumption made by platosim about the serial over-scan pixels, however why not including them ?

**Comment by Dries-Seynaeve** [comment]
It is on our to-do list. I have made a note to change the dwell time when we do implement this.
